noir-projects/noir-contracts/contracts/private_token_contract/src/main.nr

```diff
@@ -0,0 +1,229 @@
+        let to_npk_m_hash = header.get_npk_m_hash(&mut context, to);
+
+        let amount = U128::from_integer(amount);
+        storage.balances.sub(from_npk_m_hash, amount).emit(encode_and_encrypt_note_with_keys(&mut context, from_ovpk, from_ivpk));
+        storage.balances.add(to_npk_m_hash, amount).emit(encode_and_encrypt_note_with_keys(&mut context, from_ovpk, to_ivpk));
+    }
+
+    #[aztec(private)]
+    fn balance_of_private(owner: AztecAddress) -> pub Field {
+        let header = context.get_header();
+        let owner_npk_m_hash = header.get_npk_m_hash(&mut context, owner);
+        storage.balances.to_unconstrained().balance_of(owner_npk_m_hash).to_integer()
+    }
+
+    unconstrained fn balance_of_unconstrained(owner_npk_m_hash: Field) -> pub Field {
+        storage.balances.balance_of(owner_npk_m_hash).to_integer()
+    }
+
+    #[aztec(private)]
+    fn setup_refund(
+        fee_payer_npk_m_hash: Field,
+        sponsored_user: AztecAddress,
+        funded_amount: Field,
+        refund_nonce: Field
+    ) {
+        assert_current_call_valid_authwit(&mut context, sponsored_user);
+        let header = context.get_header();
+        let sponsored_user_npk_m_hash = header.get_npk_m_hash(&mut context, sponsored_user);
+        let sponsored_user_ovpk = header.get_ovpk_m(&mut context, sponsored_user);
+        let sponsored_user_ivpk = header.get_ivpk_m(&mut context, sponsored_user);
+        storage.balances.sub(sponsored_user_npk_m_hash, U128::from_integer(funded_amount)).emit(encode_and_encrypt_note_with_keys(&mut context, sponsored_user_ovpk, sponsored_user_ivpk));
+        let points = TokenNote::generate_refund_points(
+            fee_payer_npk_m_hash,
+            sponsored_user_npk_m_hash,
+            funded_amount,
+            refund_nonce
+        );
+        context.set_public_teardown_function(
+            context.this_address(),
+            FunctionSelector::from_signature("complete_refund(Field,Field,Field,Field)"),
+            [points[0].x, points[0].y, points[1].x, points[1].y]
+        );
+    }
+
+    #[aztec(public)]
+    #[aztec(internal)]
+    fn complete_refund(
+        fpc_point_x: Field,
+        fpc_point_y: Field,
+        user_point_x: Field,
+        user_point_y: Field
+    ) {
+        let fpc_point = EmbeddedCurvePoint { x: fpc_point_x, y: fpc_point_y, is_infinite: false };
+        let user_point = EmbeddedCurvePoint { x: user_point_x, y: user_point_y, is_infinite: false };
+        let tx_fee = context.transaction_fee();
+        let note_hashes = TokenNote::complete_refund(fpc_point, user_point, tx_fee);
+
+        // `compute_inner_note_hash` manually, without constructing the note
+        // `3` is the storage slot of the balances
+        context.push_new_note_hash(pedersen_hash([3, note_hashes[0]], GENERATOR_INDEX__INNER_NOTE_HASH));
+        context.push_new_note_hash(pedersen_hash([3, note_hashes[1]], GENERATOR_INDEX__INNER_NOTE_HASH));
```

Comment:
Actually no. This is purely matching `compute_inner_note_hash` in `noir-projects/aztec-nr/aztec/src/note/utils.nr`, which uses pedersen_hash, and is used in, e.g. `create_note_hash_from_public`. However, the "real" `compute_inner_note_hash` calls `compute_note_content_hash`, which is implemented on our private token using multiscalar multiplication, which is where we get the homomorphic properties. But that's why at this point we short-circuit `create_note_hash_from_public` since we effectively already have the result of `compute_inner_note_hash`.